src/app/qgsmaptooladdrectangle.cpp

```diff
@@ -71,7 +71,20 @@ void QgsMapToolAddRectangle::deactivate( )
   }
 
   mParentTool->clearCurve( );
-  mParentTool->addCurve( mRectangle.toLineString( !QgsWkbTypes::hasZ( qobject_cast<QgsVectorLayer *>( mCanvas->currentLayer() )->wkbType() ) ) );
+  // keep z value from the first snapped point
+  std::unique_ptr<QgsLineString> ls( mRectangle.toLineString() );
```

Comment:
Just for readability, maybe better:
```suggestion
  std::unique_ptr<QgsLineString> lineString( mRectangle.toLineString() );
```

Reply:
Changes pushed on the branch fix_33583. Thanks !